File: 8/ans.pl
:- dynamic(visited/1).

input([]) :- at_end_of_stream(current_input).
input([X|Xs]) :-
  read_line_to_string(current_input, Line),
  split_string(Line, " ", "", [Op, ArgString]),
  number_string(Arg, ArgString),
  X = [Op, Arg],
  input(Xs).

% Defines insn/2 dynamically
def_insn(Xs) :- def_insn(Xs, 0).
def_insn([], _).
def_insn([X|Xs], Idx) :-
  assertz(insn(Idx, X)),
  NewIdx is Idx + 1,
  def_insn(Xs, NewIdx).

interpret(["jmp", Arg], Acc, OldPC, Acc, NewPC) :- NewPC is OldPC + Arg.
interpret(["nop", _], Acc, OldPC, Acc, NewPC) :- NewPC is OldPC + 1.
interpret(["acc", Arg], OldAcc, OldPC, NewAcc, NewPC) :-
  NewAcc is OldAcc + Arg,
  NewPC is OldPC + 1.

interpret_until_repetition(Xs, Ans) :-
  def_insn(Xs),
  interpret_until_repetition(0, 0, Ans),
  retractall(insn(_, _)).
interpret_until_repetition(Acc, PC, FinalAcc) :-
  visited(PC),
  FinalAcc = Acc,
  retractall(visited(_)).
interpret_until_repetition(Acc, PC, FinalAcc) :-
  assertz(visited(PC)),
  insn(PC, Insn),
  interpret(Insn, Acc, PC, NewAcc, NewPC),
  interpret_until_repetition(NewAcc, NewPC, FinalAcc).

interpret_until_end(Xs, Ans) :-
  def_insn(Xs),
  interpret_until_end(0, 0, Ans),
  retractall(insn(_, _)).
interpret_until_end(Acc, PC, FinalAcc) :-
  insn(PC, Insn),
  interpret(Insn, Acc, PC, NewAcc, NewPC),
  interpret_until_end(NewAcc, NewPC, FinalAcc).
interpret_until_end(Acc, _PC, Acc).

replace_jmp_nop([], []).
replace_jmp_nop([["jmp", Arg]|Xs], [["nop", Arg]|Xs]).
replace_jmp_nop([["nop", Arg]|Xs], [["jmp", Arg]|Xs]).
replace_jmp_nop([X|Xs], [X|Rest]) :- replace_jmp_nop(Xs, Rest).

find_terminating_input(Xs, Ans) :-
  replace_jmp_nop(Xs, Ans),
  \+ interpret_until_repetition(Ans, _).

part2(Xs, Ans) :-
  find_terminating_input(Xs, NewXs),
  interpret_until_end(NewXs, Ans).

main :-
  input(Xs),
  interpret_until_repetition(Xs, Ans1),
  writeln(Ans1),
  part2(Xs, Ans2),
  writeln(Ans2).

File: 8/ans.zig
const std = @import("std");
const stdin = std.io.getStdIn().reader();
const stdout = std.io.getStdOut().writer();
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;

const InsnType = enum {
    jmp,
    nop,
    acc,
};
const Insn = struct { insnType: InsnType, arg: i64 };

const MAX_INPUT_SIZE = 1024;

fn readInput(allocator: *Allocator) !ArrayList(Insn) {
    var buf: [MAX_INPUT_SIZE]u8 = undefined;

    var list = ArrayList(Insn).init(allocator);
    while (true) {
        const line = (try stdin.readUntilDelimiterOrEof(&buf, '\n')) orelse return list;
        var iterator = std.mem.split(line, " ");
        var insnTypeString = iterator.next().?;
        var argString = iterator.next().?;

        const insnType = std.meta.stringToEnum(InsnType, insnTypeString).?;
        const arg = try std.fmt.parseInt(i64, argString, 10);

        try list.append(.{ .insnType = insnType, .arg = arg });
    }
}

fn part1(input: anytype) !i64 {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = &arena.allocator;

    var visited = std.AutoHashMap(i64, void).init(allocator);
    var acc: i64 = 0;
    var pc: i64 = 0;
    while (true) {
        if (visited.contains(pc)) {
            return acc;
        }
        try visited.putNoClobber(pc, {});
        const insn = input.items[@intCast(usize, pc)];
        switch (insn.insnType) {
            InsnType.jmp => {
                pc += insn.arg;
            },
            InsnType.nop => {
                pc += 1;
            },
            InsnType.acc => {
                pc += 1;
                acc += insn.arg;
            },
        }
    }
}

fn terminates(input: anytype) !?i64 {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = &arena.allocator;

    var visited = std.AutoHashMap(i64, void).init(allocator);
    var acc: i64 = 0;
    var pc: i64 = 0;
    while (true) {
        if (visited.contains(pc)) {
            return null;
        }
        if (pc >= input.items.len) {
            return acc;
        }
        try visited.putNoClobber(pc, {});
        const insn = input.items[@intCast(usize, pc)];
        switch (insn.insnType) {
            InsnType.jmp => {
                pc += insn.arg;
            },
            InsnType.nop => {
                pc += 1;
            },
            InsnType.acc => {
                pc += 1;
                acc += insn.arg;
            },
        }
    }
}

fn part2(input: anytype) !?i64 {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = &arena.allocator;

    for (input.items) |*insn, i| {
        const originalInsnType = insn.insnType;
        const newInsnType = switch (insn.insnType) {
            InsnType.acc => continue,
            InsnType.nop => InsnType.jmp,
            InsnType.jmp => InsnType.nop,
        };
        const newInsn = Insn{ .insnType = newInsnType, .arg = insn.arg };
        insn.* = newInsn;
        if (try terminates(input)) |acc| {
            return acc;
        }
        insn.* = Insn{ .insnType = originalInsnType, .arg = insn.arg };
    }
    return null;
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = &arena.allocator;

    const input = try readInput(allocator);
    const ans1 = try part1(input);
    try stdout.print("{}\n", .{ans1});
    const ans2 = (try part2(input)).?;
    try stdout.print("{}\n", .{ans2});
}
